### PR TITLE
VCH restart for networking [full ci]

### LINKED ISSUE #2598: Test teardown failed in CI
Seeing this failure in several builds:
```
Test-Cases.Group1-Docker-Commands.1-1-Docker-Info :: Test 1-1 - Do... | FAIL |
Suite teardown failed:
Several failures occurred:

1) '/ha-datacenter/vm/VCH-5447-7793
/ha-datacenter/vm/trusting_sinous-0e852fdc76ede40ec8fb977e7ecc006df8130d421a68ed2e1499c1aef9427c5c' contains 'VCH-5447-7793'

2) 'trusting_sinous-0e852fdc76ede40ec8fb977e7ecc006df8130d421a68ed2e1499c1aef9427c5c
test
VCH-5447-7793' contains 'VCH-5447-7793'

3) '/ha-datacenter/host/vic-esx-09.ci.drone.local/Resources/VCH-5447-7793' contains 'VCH-5447-7793'

4) 1 != 0

5) 'INFO[2016-10-06T18:51:20Z] ### Removing VCH ####                        
INFO[2016-10-06T18:51:20Z]                                              
INFO[2016-10-06T18:51:20Z] VCH ID: VirtualMachine:1493                  
INFO[2016-10-06T18:51:20Z] Removing VMs                                 
ERRO[2016-10-06T18:51:21Z] Failed to invoke task operation : Another task is already in progress. 
INFO[2016-10-06T18:51:21Z] Specify --force to force delete              
INFO[2016-10-06T18:51:21Z] Collecting ha-host hostd.log                 
ERRO[2016-10-06T18:51:21Z] Failed to destroy VM "VirtualMachine:1494": Another task is already in progress. then failed to unregister VM: ServerFaultCode: Another task is already in progress. 
ERRO[2016-10-06T18:51:21Z] --------------------                         
ERRO[2016-10-06T18:51:21Z] vic-machine-linux failed: delete failed
     ' does not contain 'Completed successfully'
```
[Uploading integration_logs_5447_5b0b864e80be5a6ab1ee59d71026dd8dd9588b15.zip…]()

### PR DESCRIPTION
**Pull request by hmahmood**
Includes:

* Rebuilding network layer objects from containers VM configs that are present in the VCH's resource pool. The big assumption here is that the state of the containers does not change during recreation of the various network objects, with one exception: running containers that stop during recreation. The latter case will be handled by the event callback from vSphere.
* General fixes for data races in exec and network layers
* Fix for container refresh code when listing scopes. This is only done for containers that are connected to a scope that is not managed by the network layer, i.e. containers that are connected to networks with DHCP.

Fixes #2598

### CODE REVIEW

**Comment by hmahmood** [comment]
@caglar10ur @hickeng

File: lib/apiservers/portlayer/restapi/handlers/scopes_handlers.go [hunk 1]
```diff
@@ -88,47 +85,16 @@ func parseScopeConfig(cfg *models.ScopeConfig) (subnet *net.IPNet, gateway net.I
 
 func (handler *ScopesHandlersImpl) listScopes(idName string) ([]*models.ScopeConfig, error) {
 	defer trace.End(trace.Begin(idName))
-	_scopes, err := handler.netCtx.Scopes(&idName)
+	_scopes, err := handler.netCtx.Scopes(context.Background(), &idName)
```
Comment (fdawg4l) [inline lib/apiservers/portlayer/restapi/handlers/scopes_handlers.go]: `_scopes`?  Is that a python'ism?

File: lib/config/virtual_container_host.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ type VirtualContainerHostConfigSpec struct {
 	BridgeNetwork       string `vic:"0.1" scope:"read-only" key:"bridge_network"`
 	CreateBridgeNetwork bool   `vic:"0.1" scope:"read-only" key:"create_bridge_network"`
 	// Published networks available for containers to join, keyed by consumption name
```
Comment (fdawg4l) [inline lib/config/virtual_container_host.go]: What is a "consumption name"?

File: lib/portlayer/exec/container.go [hunk 3]
```diff
@@ -150,33 +148,45 @@ func (s State) String() string {
...
-func (c *Container) NewHandle() *Handle {
+func (c *Container) NewHandle(ctx context.Context) *Handle {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// Call property collector to fill the data
+	if c.vm != nil {
+		if err := c.refresh(ctx); err != nil {
+			log.Errorf("refreshing container %s failed: %s", c.ExecConfig.ID, err)
+			return nil
```
Comment (fdawg4l) [inline lib/portlayer/exec/container.go]: Returning `nil` on error.  Can you add a comment as to why this is ok?

File: lib/portlayer/network/network.go [hunk 4]
```diff
@@ -48,8 +54,14 @@ func (e DuplicateResourceError) Error() string {
 }
 
 func Init(ctx context.Context, sess *session.Session, source extraconfig.DataSource, sink extraconfig.DataSink) error {
-	var err error
-	initializer.Do(func() {
+	trace.End(trace.Begin(""))
+
+	initializer.once.Do(func() {
+		var err error
```
Comment (fdawg4l) [inline lib/portlayer/network/network.go]: There is a scoped error on line 79.  It doesn't look like a problem, but if you were to rely on this defer func to capture your err in side the for loop defined in on line 73 without seeing the shadow declaration on 79, that would be bad.  May want to rename that error on 79.

File: lib/config/virtual_container_host.go (hunk 2 above)
Comment (hmahmood) [inline lib/config/virtual_container_host.go]: The name under which they show up in docker.

File: lib/apiservers/portlayer/restapi/handlers/containers_handlers.go [hunk 5]
```diff
@@ -201,7 +201,8 @@ func (handler *ContainersHandlersImpl) GetStateHandler(params containers.GetStat
 func (handler *ContainersHandlersImpl) GetHandler(params containers.GetParams) middleware.Responder {
 	defer trace.End(trace.Begin(params.ID))
 
-	h := exec.GetContainer(uid.Parse(params.ID))
+	ctx := context.Background()
+	h := exec.GetContainer(ctx, uid.Parse(params.ID))
```
Comment (caglar10ur) [inline lib/apiservers/portlayer/restapi/handlers/containers_handlers.go]: [cosmetic] 

```h := exec.GetContainer(context.Background(), uid.Parse(params.ID))```

as that's what you are doing below :)

File: lib/config/executor/network_interface.go [hunk 6]
```diff
@@ -29,8 +29,11 @@ type NetworkEndpoint struct {
 	// Common.ID - pci slot of the vnic allowing for interface identifcation in-guest
 	Common
 
-	// IP address to assign - nil if DHCP
-	Static *net.IPNet `vic:"0.1" scope:"read-only" key:"staticip"`
+	// Whether this endpoint's IP was specified by the client (true if it was)
+	Static bool `vic:"0.1" scope:"read-only" key:"static"`
+
+	// IP address to assign
+	IP *net.IPNet `vic:"0.1" scope:"read-only" key:"eip"`
```
Comment (caglar10ur) [inline lib/config/executor/network_interface.go]: should we change the key to ip here and make the next one assigned?

File: lib/portlayer/exec/container.go [hunk 7]
```diff
@@ -150,33 +148,45 @@ func (s State) String() string {
...
 	var o mo.VirtualMachine
 
 	// make sure we have vm
 	if c.vm == nil {
 		return fmt.Errorf("There is no backing VirtualMachine %#v", c)
 	}
-	ctx := context.TODO()
-
-	if err := c.vm.Properties(ctx, c.vm.Reference(), []string{"config", "runtime"}, &o); err != nil {
+	if err := c.vm.Properties(ctx, c.vm.Reference(), []string{"config", "config.extraConfig", "runtime"}, &o); err != nil {
```
Comment (caglar10ur) [inline lib/portlayer/exec/container.go]: config should be returning config.extraConfig as well, is it not?

File: lib/portlayer/network/context.go [hunk 8]
```diff
@@ -68,59 +67,71 @@ type AddContainerOptions struct {
...
 	if config == nil {
 		return nil, fmt.Errorf("missing config")
 	}
 
-	pones, pbits := bridgePool.Mask.Size()
-	mones, mbits := bridgeMask.Size()
+	bridgeRange := config.BridgeIPRange
+	if bridgeRange == nil || len(bridgeRange.IP) == 0 || bridgeRange.IP.IsUnspecified() {
+		var err error
+		_, bridgeRange, err = net.ParseCIDR("172.16.0.0/12")
```
Comment (caglar10ur) [inline lib/portlayer/network/context.go]: should we make this a constant and move into constants pkg?

File: lib/portlayer/network/network.go [hunk 9]
```diff
@@ -64,60 +76,131 @@ func Init(ctx context.Context, sess *session.Session, source extraconfig.DataSou
...
-			if _, err = netctx.BindContainer(h); err != nil {
-				return
-			}
+// handleEvent processes events
+func handleEvent(netctx *Context, ie events.Event) {
+	switch ie.String() {
+	case events.ContainerPoweredOff:
+		handle := exec.GetContainer(context.Background(), uid.Parse(ie.Reference()))
+		if handle == nil {
+			log.Errorf("Container %s not found - unable to UnbindContainer", ie.Reference())
```
Comment (caglar10ur) [inline lib/portlayer/network/network.go]: any reason why we are logging an error but not doing anything?

File: lib/tether/config.go [hunk 10]
```diff
@@ -103,8 +104,11 @@ type NetworkEndpoint struct {
 	// Common.ID - pci slot of the vnic allowing for interface identifcation in-guest
 	executor.Common
 
-	// IP address to assign - nil if DHCP
-	Static *net.IPNet `vic:"0.1" scope:"read-only" key:"staticip"`
+	// Whether this endpoint's IP was specified by the client (true if it was)
+	Static bool `vic:"0.1" scope:"read-only" key:"static"`
+
+	// IP address to assign
+	IP *net.IPNet `vic:"0.1" scope:"read-only" key:"eip"`
```
Comment (caglar10ur) [inline lib/tether/config.go]: same comment, should we rename the keys?

File: pkg/vsphere/extraconfig/decode.go [hunk 11]
```diff
@@ -266,7 +266,7 @@ func decodeSlice(src DataSource, dest reflect.Value, prefix string, depth recurs
 	}
 
 	var this reflect.Value
-	if !dest.IsValid() || dest.IsNil() {
+	if !dest.IsValid() || dest.IsNil() || length > dest.Cap() {
```
Comment (caglar10ur) [inline pkg/vsphere/extraconfig/decode.go]: I think this is already in a different PR

Comment (hmahmood) [inline pkg/vsphere/extraconfig/decode.go]: Yes this should go away once I rebase.

File: lib/tether/config.go (hunk 10 above)
Comment (hmahmood) [inline lib/tether/config.go]: Yes I renamed them as you suggested.

**Comment by caglar10ur** [comment]
LGTM

**Comment by andrewtchin** [comment]
lgtm

File: tests/test-cases/Group10-VCH-Restart/10-1-VCH-Restart.robot [hunk 12]
```diff
@@ -40,13 +56,8 @@ Created Network And Images Persists As Well As Containers Are Discovered With Co
...
 
-    Log To Console  Rebooting VCH ...
-    ${rc}  ${output}=  Run And Return Rc And Output  govc vm.power -reset=true ${vch-name}
-    Should Be Equal As Integers  ${rc}  0
-    Log To Console  Waiting for VCH to boot ...
-    Wait Until Vm Powers On  ${vch-name}
-    Log To Console  VCH Powered On
-    Sleep  5
+    Reboot VCH
+    Sleep  10
```
Comment (mhagen-vmware) [inline tests/test-cases/Group10-VCH-Restart/10-1-VCH-Restart.robot]: Do we not want to use at least a bit of the wait until vm powers on?  I think ideally, in these cases, we could wait until vm powered on, then wait until docker info succeeds?

**Comment by hmahmood** [comment]
♻️

**Pull request merged**